File: CMPS3500_MiniScheme_Starter_Package/procedural/let.cpp
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga  /
  / ASGT: CHECKPOINT 2                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: let.cpp                                  /
  / DATE: 04/11/2026                                    /
*/

#include "let.h"
#include "evaluate.h"
#include "scope.h"

// Handles the built-in let expression
std::string handleLet(const std::vector<std::string>& expr, Scope* scope)
{
    Scope* new_scope = enterScope(scope);

    int i = 1;

    if (expr[i] != "(") return "ERROR";
    i++;

    while (expr[i] == "(")
    {
        i++;

        std::string var = expr[i];
        i++;

        std::vector<std::string> value_expr;

        while (expr[i] != ")")
        {
            value_expr.push_back(expr[i]);
            i++;
        }

        i++;

        std::string value = evaluate(value_expr, scope);
        addScopeEntry(new_scope, var, value);
    }

    
    if (expr[i] != ")") return "ERROR";
    i++;

    std::vector<std::string> body;

    // Extract the body of the let expression
    while (i < (int)expr.size())
    {
        body.push_back(expr[i]);
        i++;
    }

    std::string result = evaluate(body, new_scope);

    exitScope(new_scope);

    return result;
}

File: CMPS3500_MiniScheme_Starter_Package/oop/Let.java
/*
  / NAME: Henry Arinaga, Alberto Molina, Peter Uzuriaga /
  / ASGT: CHECKPOINT 3                                  /
  / ORGN: CSUB - CMPS 3500                              /
  / FILE: Let.java                                      /
  / DATE: 05/01/2026                                    /
*/

import java.util.List;
import java.util.ArrayList;

public class Let extends ExpressionHandler {
    public Let(List<String> expression){
        super(expression, 1);
    }

    public static String handleLet(List<String> expression, Scope scope){
        Let letExpression = new Let(expression);
        Scope newScope = scope.enterScope();

        if(letExpression.index >= expression.size() || !expression.get(letExpression.index).equals("(")){
            return "PARSE_ERROR";
        }

        letExpression.index++;

        while(letExpression.index < expression.size() && expression.get(letExpression.index).equals("(")){
            letExpression.index++;

            if(letExpression.index >= expression.size()){
                newScope.exitScope();
                return "PARSE_ERROR";
            }

            String var = expression.get(letExpression.index);
            letExpression.index++;

            List<String> valueExpression = letExpression.extractPart();

            if(valueExpression.isEmpty() || letExpression.index >= expression.size()){
                newScope.exitScope();
                return "PARSE_ERROR";
            }

            letExpression.index++;

            String value = Evaluate.evaluate(valueExpression, newScope);

            if(getError(value)){
                newScope.exitScope();
                return "PARSE_ERROR";
            }

            newScope.addScopeEntry(var, value);
        }

        if(letExpression.index >= expression.size() || !expression.get(letExpression.index).equals(")")){
                newScope.exitScope();
                return "PARSE_ERROR";
        }

        letExpression.index++;

        List<String> body = new ArrayList<>();

        while(letExpression.index < expression.size()){
            body.add(expression.get(letExpression.index));
            letExpression.index++;
        }

        String result = Evaluate.evaluate(body, newScope);
        newScope.exitScope();
        return result;
    }
}
